cmd/testscript/main.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2018 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/rogpeppe/go-internal/goproxytest"
+	"github.com/rogpeppe/go-internal/gotooltest"
+	"github.com/rogpeppe/go-internal/testscript"
+	"github.com/rogpeppe/go-internal/txtar"
+)
+
+var (
+	fVerbose = flag.Bool("v", false, "run tests verbosely")
+
+	runDir string
+)
+
+func main() {
+	os.Exit(main1())
+}
+
+func main1() int {
+	if err := mainerr(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return 1
+	}
+	return 0
+}
+
+func mainerr() (retErr error) {
+	flag.Usage = func() {
+		mainUsage(os.Stderr)
+	}
+	flag.Parse()
+
+	td, err := ioutil.TempDir("", "testscript_")
+	if err != nil {
+		return fmt.Errorf("unable to create temp dir: %v", err)
+	}
+	defer os.RemoveAll(td)
+	runDir = td
+
+	files := flag.Args()
+	if len(files) == 0 {
+		files = []string{os.Stdin.Name()}
+	}
+
+	for i, fn := range files {
+		// TODO make running files concurrent by default? If we do, note we'll need to do
+		// something smarter with the runner stdout and stderr below
+		rd := filepath.Join(td, "dir"+strconv.Itoa(i))
+		if err := os.Mkdir(rd, 0777); err != nil {
+			return fmt.Errorf("failed to create a run directory within %v for %v: %v", td, fn, err)
+		}
+		if err := run(rd, fn); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+var failedRun = errors.New("failed run")
+
+type runner struct {
+	verbose bool
+}
+
+func (r runner) Skip(is ...interface{}) {
+	panic("Skip is not supported")
+}
+
+func (r runner) Fatal(is ...interface{}) {
+	r.Log(is...)
+	r.FailNow()
+}
+
+func (r runner) Parallel() {
+	// No-op for now; we are currently only running a single script in a
+	// testscript instance.
+}
+
+func (r runner) Log(is ...interface{}) {
+	fmt.Print(is...)
+}
+
+func (r runner) FailNow() {
+	panic(failedRun)
+}
+
+func (r runner) Run(n string, f func(t testscript.T)) {
+	// For now we we don't top/tail the run of a subtest. We are currently only
+	// running a single script in a testscript instance, which means that we
+	// will only have a single subtest.
+	f(r)
+}
+
+func (r runner) Verbose() bool {
+	return r.verbose
+}
+
+func run(rd, fn string) error {
+	var ar *txtar.Archive
+	var err error
+
+	testdata := filepath.Join(rd, "testdata")
+	mods := filepath.Join(testdata, "goproxytest")
+
+	if err := os.MkdirAll(mods, 0777); err != nil {
+		return fmt.Errorf("failed to create testdata dir: %v", err)
+	}
+
+	if fn == os.Stdin.Name() {
```

Comment:
I'm not sure about this. On linux this will happen automatically (opening `/dev/stdin` reads stdin). On windows, who knows what os.Stdin.Name() returns? I'd either go with a standard name (`-`) or just say "if there are no files on the command line, read from stdin".

Reply:
I'll go with the standard name, thanks.